Logging

samansmink opened the pull request:
This draft PR introduces a Logger to DuckDB. The logger will enable the communication of various types of information to both users and developers.

The logger works without any globals and should be fast enough to have log statements compiled into the release builds. By default logging will be disabled but should be configurable through a runtime global setting

## How to use
First of all, the logger currently has 6 levels: `TRACE`, `DEBUG`, `INFO`, `WARN`, `ERROR` and `FATAL`.

Secondly, the logger uses no globals, a log statement always requires some sort of context. We currently support `ThreadContext`, `ExecutionContext`, `ClientContext`, `FileOpener` and `DatabaseInstance`. 

### Basics
This PR includes a bunch of templating magic to make this work, so our basic log statement looks like:
```C++
Logger::Warn(db_instance, "hi i'm a log entry");
```
We can also use other log levels or contexts:
```C++
Logger::Info(client_context, "this goes to the client_context local logger");
Logger::Fatal(thread_context, "this goes to the thread local logger");
```

### String construction
Because log statements are intended for release binaries too, we should ensure that we can cheaply run the log statements without doing any more work than necessary when logging is disabled. For this reason I have also added format strings log statements:
```C++
Logger::Warn(db, "look at my pretty log: '%s'", some_string);
```
 and callback based log statements:
```C++
Logger::Info(db, [](){
	return SomeExpensiveStringConstructionFunction();
});
```

### Enabling logging
Logging is disabled by default, but can be enabled using
```SQL
set enable_logging=true;
set logging_level='warn';
```

### Log types
Log entries have a type that is currently set to "default" everywhere. However you can specify a custom type to be set using:

```C++
Logger::Info("my_log_type", client_context, "this goes to the client_context local logger");
```

these types can be used to filter using an inclusion list:
```SQL
set enable_logging=true;
set enabled_loggers='my_log_type1,my_log_type2'
set logging_mode='enable_selected'
```

or an exclusions list:
```SQL
set disable_logging=true;
set disabled_loggers='my_log_type1,my_log_type2'
set logging_mode='disable_selected'
```

### Log output
In the current draft state the log entries are written to a ColumnDataCollection stored in the LogManager in the Databaseinstance. We should however be able to switch between that, writing to stdout directly, and appending to a file.

## Implementation
The main driver behind this implementation is that log settings should be configurable at runtime. The consequence of this is that we need to take special care to avoid needing to lock some global state or check atomics on every log statement. This is achieved by caching the logging configuration at various points.

### 3 Layers of logging
Currently, there are 3 places where loggers are stored:
- In the LogManager (which lives in the DatabaseInstance)
	- uses atomics to check if logging is enabled
- In the ClientContext 
	- uses const values to check if logging is enabled
	- refreshed on every QueryStart and QueryEnd
- In the ThreadContext
	- uses const values to check if logging is enabled

Besides caching the logging config, these loggers can also store additional context information such as the thread_id and the transaction_id.

Also, the logger in the ThreadContext, is thread-local, which we can use later to optimize performance since it can buffer log entries in a thread-local cache, wheres the higher level loggers require a global lock for writing log entries.

## Performance
I've added some benchmarks that are also added to the regression tester to ensure we aren't introducing any measurable slowdown by leaving in log statements in release binaries. The most important benchmarks are `benchmark/micro/logger/disabled/*.benchmark` since these measure the overhead of a Log statement when logging is disabled. The results on my M1 Max macbook pro are:

|                                    name                                    |       result       | relative_to_reference |
|----------------------------------------------------------------------------|-------------------:|----------------------:|
| benchmark/micro/logger/disabled/logging_disabled_global.benchmark          | 0.926281          | 2.210263585898307     |
| benchmark/micro/logger/disabled/logging_disabled_reference.benchmark       | 0.419082          | 1.0                   |
| benchmark/micro/logger/disabled/logging_disabled_client_context.benchmark  | 0.60807 | 1.4509584768216974    |
| benchmark/micro/logger/disabled/logging_disabled_thread_local.benchmark    | 0.60227 | 1.4371349159686324    |
| benchmark/micro/logger/disabled/logging_disabled_file_opener.benchmark     | 0.64255 | 1.533238872811405     |

In these benchmarks, a single-threaded table function is called which runs a `range()` table function that generates a single INT column with values from 0 to 50000000 in steps of 1. In the inner loop of this range function, the logger is called once for each tuple. The for the `logging_disabled_reference`  benchmark, the log statement is not called at all serving as a reference for the other benchmarks. The results show that the Log call is cheap, but also not free when logging is disabled. Furthermore, we can see that the global logger is the most expensive, which makes sense as it needs to use atomics instead of const values. The FileOpener logger is also slightly more expensive which is caused by one extra pointer that needs to be followed for checking if logging is enabled.

The conclusion here is that the Log calls are very cheap and can be left in the code basically everywhere except for the performance critical tight inner loops. 

## Log Type naming convention
Log entries can contain a `log_type`. I propose the following namespaced naming scheme for the log types to keep them sane:
`<codebase>.<NameSpaceName>.<NameSpaceName>...<TypeName>`

for `<codebase>`, this will be any of:
- `duckdb` for logs coming from duckdb itself
- `<extension_name>` for logs coming from extensions
- `duckdb-<client>` for logs coming from duckdb client code (python, node, etc)

For example, I have currently added:
`duckdb.ClientContext.BeginQuery`, `duckdb.FileSystem.LocalFileSystem.OpenFile`, `duckdb.Extensions.ExtensionLoaded`, and `duckdb.Extensions.ExtensionAutoLoaded`

## Remaining work
There's a few things still remaining, though these may potentially be split across follow up PRs:
- Make LogStorage pluggable for extensions
- determine guidelines for default logging behaviour in duckdb/duckdb and core extensions
- add C API calls 

### Prefix matching Log Types
Currently the log types are added to a set and then (inefficiently) string matched when the LogMode is set to `DISABLE_SELECTED` or `ENABLE_SELECTED`. However, with namespaced log type names, what we would actually like is to be able to do prefix matching. For example in:
```SQL
set enabled_loggers='duckdb.FileSystem,httpfs.FileSystem,duckdb.Extensions'
set logging_mode='enable_selected'
```
we would like to have DuckDB log all entries with log types that start with one of these prefixes.

there are basically 2 ways I see of doing this. 1 is to use a trie, perhaps using our own ART. I have discussed this with @taniabogatsch and we have concluded that we would need to make some changes to the ART to make it lightweight enough to fulfil this task. The other approach would be to create some sort of mapping where log types are pre-registered and then passed to the log calls as numericals. The latter would be the fastest for sure, but would also require registering log types separately and keeping some sort of state

samansmink committed:
loggin wip

samansmink committed:
wip logging

samansmink committed:
first basic prototype of logger working-ish

samansmink committed:
Logging API is done?

samansmink committed:
missing file

samansmink committed:
add client context

samansmink committed:
add timestamps

samansmink committed:
more logging stuff

samansmink committed:
wip testing logging

samansmink committed:
add proper logging test

samansmink committed:
fix various issues discovered by exhaustive tests

samansmink committed:
some small touches to logging

Mytherin commented on src/logging/logger.cpp:
```diff
@@ -0,0 +1,437 @@
+}
+
+Logger& Logger::Get(FileOpener &opener) {
+	auto maybe_context = opener.TryGetClientContext();
```
Maybe this should be inside the file opener so we don't need to branch here

samansmink committed:
split up logger into files

samansmink committed:
basic stdout logging

samansmink committed:
make combined function through bind_replace

samansmink committed:
fix FileOpener logging

samansmink committed:
wip

samansmink committed:
add benchmarking for logging

samansmink committed:
add benchmarks for all loggers

samansmink committed:
logger cleanup

samansmink committed:
format

samansmink committed:
avoid logger outliving buffer manager

samansmink committed:
wip locking scans

samansmink committed:
fix issue with buffer managed log entries

samansmink committed:
fix context registering issue

samansmink committed:
format

samansmink committed:
increase buffer size for log storage

samansmink committed:
add slow test for many log entries

samansmink committed:
cleanup logger

samansmink committed:
add trace log level, rename debugging to debug

samansmink committed:
Merge branch 'main' into logging-wip

samansmink committed:
rename enableloggers disableloggers

samansmink committed:
add missing logging benchmarks to ci

samansmink committed:
use naming convention for log

samansmink committed:
format

samansmink committed:
naming consistency

samansmink committed:
cleanup

samansmink committed:
add log type filter message

samansmink committed:
rip out query log

samansmink committed:
add reset test options, fix small bug

samansmink committed:
fix indeterministic test

samansmink committed:
fix outdated test

samansmink committed:
ci fixes

samansmink committed:
fix syntax error on msvc

samansmink committed:
jump the make tidy hoops

samansmink committed:
missing typedefs

samansmink committed:
format

samansmink committed:
add missing header

samansmink committed:
fix small compiler complaints

samansmink committed:
disable tests for force restart ci

samansmink committed:
remove function to avoid amalgamation issues

samansmink committed:
remove accidentally added header

samansmink committed:
fix log level enum conflict with windows macro

hannes commented:
Thanks! I'm a bit wary about the log types - it seems to rely a lot on consistent behavior by all contributors. it would be difficult to turn on logging on something without actually reading the code. It is also non-obvious which logging group a log entry belongs to, it could be multiple. Maybe we should figure out if there is a better way of doing this, e.g. defaulting on class name, file name, file path etc. The code formatter could enforce the convention possibly.

hannes commented on src/common/enum_util.cpp:
```diff
@@ -2002,6 +2003,66 @@ LoadType EnumUtil::FromString<LoadType>(const char *value) {
+
+const StringUtil::EnumStringLiteral *GetLogLevelValues() {
+	static constexpr StringUtil::EnumStringLiteral values[] {
+		{ static_cast<uint32_t>(LogLevel::TRACE), "TRACE" },
```
maybe a prefix like `L_` would be nicer so its consistent and obvious.

hannes commented on src/function/table/system/duckdb_log.cpp:
```diff
@@ -0,0 +1,115 @@
+		return_types.emplace_back(LogicalType::VARCHAR);
+
+		names.emplace_back("client_context");
+		return_types.emplace_back(LogicalType::UBIGINT);
```
There was an issue with `UBIGINT` being unwieldy when looking at it in the shell, please check

hannes commented on src/function/table/system_functions.cpp:
```diff
@@ -23,6 +23,9 @@ void BuiltinFunctions::RegisterSQLiteFunctions() {
 	DuckDBDatabasesFun::RegisterFunction(*this);
 	DuckDBFunctionsFun::RegisterFunction(*this);
 	DuckDBKeywordsFun::RegisterFunction(*this);
+	TestLoggingFun::RegisterFunction(*this);
```
maybe this could be a test-only extension like the `loadable_extension_test` or so

hannes commented on src/include/duckdb/logging/logging.hpp:
```diff
@@ -0,0 +1,80 @@
+
+namespace duckdb {
+
+enum class LogLevel : uint8_t {
```
I understand the idea behind log levels, but its not clear what a `FATAL` would even be in DuckDB world, or what the difference between `TRACE` and `DEBUG` should be

samansmink committed:
rename log enum

rcoup commented:
@samansmink is it worth (long-term) wiring this into Python, Java, and other language bindings native/standard logging systems so messages propagate through them rather than being managed/output directly by DuckDB? 

eg: so a python user can do something like
```
import logging.config

logging.config.dictConfig({
    ...
    'loggers': {
        'duckdb.sql': {
            'level': 'WARNING',
        },
        'duckdb.spatial': {
            'level': 'DEBUG',
        },
    },
    'root': {
        'level': 'INFO',
        'handlers': ['stderr'],
    }
})
```

From a quick look maybe that can be handled via a `logging_storage='client'` or similar? Maybe that should eventually be the default.

Mytherin commented on benchmark/micro/logger/filtered_out_by_log_type/thread_local.benchmark:
```diff
@@ -0,0 +1,20 @@
+
+# Note: this will call the Logger, but the log type is filtered out
+run
+SELECT sum(value) FROM test_logging(50000000, enable_logging=true, log_message='hello world', logger='thread_local', log_type='duckdb.SomeName.SomeOtherName.BlaBla');
```
This function can e.g. return `NULL` by default, and take an optional `return_value`, e.g.:

```sql
SELECT log_value(printf('value of i: %d', i), return_value := i)
```

Mytherin commented on src/function/table/system/duckdb_log.cpp:
```diff
@@ -0,0 +1,115 @@
+	if (!context_id_only) {
+		if (LogManager::Get(context).CanScan()) {
+			string query =
+			    "SELECT * exclude (l.context_id, c.context_id) FROM duckdb_logs(context_id_only=true) as l JOIN "
```
Can this be made into a view instead (in `default_views.cpp`)

Mytherin commented on src/include/duckdb/logging/logging.hpp:
```diff
@@ -0,0 +1,80 @@
+
+enum class LogLevel : uint8_t {
+	TRACE = 10,
+	DEBUGGING = 20, // ToString-ed as DEBUG, but named DEBUGGING due to conflict with DEBUG define
```
Maybe prefix all of these with `LOG`, e.g. `LOG_DEBUG`

Mytherin commented on src/include/duckdb/main/settings.hpp:
```diff
@@ -763,14 +793,33 @@ struct LockConfigurationSetting {
-struct LogQueryPathSetting {
+struct LoggingLevel {
 	using RETURN_TYPE = string;
-	static constexpr const char *Name = "log_query_path";
```
Can we keep this setting and then turn on the logger to log similarly to the old query path logger?

Mytherin replied:
Or for now we can leave the old logger in-tact and leave replacing it for a future PR

samansmink replied:
I've restored it for now, will remove again when I add the file log storage

Mytherin commented on src/logging/log_manager.cpp:
```diff
@@ -0,0 +1,146 @@
+	LogConfig config_copy;
+	{
+		unique_lock<mutex> lck(lock);
+		config_copy = config;
```
Do we need to make this copy?

samansmink replied:
I remember the reason now: in the constructor of a Logger, a Logger needs to register itself to the LogManager which requires locking the LogManager. See for example `ThreadSafeLogger(LogConfig &config_p, LoggingContext &context_p, LogManager &manager)`.  I did this to avoid deadlocking. However this is indeed not super clean.

I will just add another constructor that takes a RegisteredLogContext instead

Mytherin reviewed (commented):
Thanks for the fixes! Looks great - some more minor comments:

samansmink committed:
wip adding scalar function log write_log

samansmink committed:
fix issue where optimizer eats expression aliases

samansmink committed:
switch test logging function from table function to scalar

This scalar function is now generally useful instead of some bogus testing function

samansmink committed:
remove config copy in CreateLogger

samansmink committed:
change log level enum prefix once more

samansmink committed:
remove bindreplace in favor of doing join in view

samansmink committed:
restore query log for now

samansmink committed:
make log storage pluggable

samansmink committed:
Merge branch 'main' into logging-wip

samansmink committed:
cleanup new logger code

samansmink commented:
### Rewrite test function to scalar
Slightly bigger fix was to rewrite the test function for logging to a ScalarFunction. This makes the function generally useful because it can also be used to write to the log through SQL, for example:

```SQL
SELECT write_log('this is my log message', level := 'warn', scope := 'database', log_type := 'some.log.type' )
```

Note that the return value of write_log is NULL by default, but it can be defined using the return value parameter:
```SQL
SELECT write_log('logging: ' || i::VARCHAR, return_value := i) as i from range(0,10) tbl(i)
```

I've rewritten the benchmarks and tests to use this function now. 

### Pluggable LogStorage
The logstorage is now pluggable, see `test/logging/test_logging.cpp`:
```C++
auto log_storage = make_shared_ptr<MyLogStorage>();
db.instance->GetLogManager().RegisterLogStorage("my_log_storage", log_storage);
```
This may require some more thought later, but the basics should work


### Minor fixes
- LogLevel Enum is now prefixed with `LOG_`. The prefix is removed for `EnumUtil::ToString`
- the join between the log_entries and log_contexts is now done in the view, not in a bindreplace
- query log is restored for now, will remove once file log is added
- remove weird config copy

samansmink committed:
clang tidy fixes

samansmink committed:
prevent overwriting expression alias

samansmink committed:
tidy fixes

samansmink committed:
fix include

samansmink commented:
@Mytherin I think this is one is good to go, failure is a regression test that fails on the old run because the feature is new

Mytherin commented:
Thanks!

carlopi commented on src/logging/log_manager.cpp:
```diff
@@ -0,0 +1,155 @@
+		throw NotImplementedException("File log storage is not yet implemented");
+	} else if (registered_log_storages.find(storage_name_to_lower) != registered_log_storages.end()) {
+		log_storage = registered_log_storages[storage_name_to_lower];
+	}
```
@samansmink: should there be a:
```
else {
    throw NotImplementedException("Log Storage %s not implemented and not registered", storage_name_to_lower);
}
```

Currently:
```
SET logging_storage='abc';
```
succeeds but do not actually change `log_storage`.
